Add C8 REST guidelines

tmetzke opened the pull request:
## Description

* Add a public version of the [C8 REST API guidelines](https://docs.google.com/document/d/1G9AmmNac-4QLGZ0LXQXa3FyeCrSJdIboaPt3-R6dWNw) to help users understand how the C8 REST API works and can be navigated.
* Backport the guide to 8.6 to provide general guidance (only those parts present there).
* Sort the 8.6 C8 REST API spec categories for better usability.

closes https://github.com/camunda/camunda/issues/24239

## When should this change go live?

<!-- PRs merged go to stage.docs.camunda.io first and must be manually released to docs.camunda.io. -->

- [ ] This is a bug fix, security concern, or something that needs **urgent release support**.
- [ ] This is already available but undocumented and should be released within a week.
- [ ] This on a **specific schedule** and the assignee will coordinate a release with the DevEx team. (apply `hold` label or convert to draft PR)
- [ ] This is part of a scheduled alpha or minor. (apply alpha or minor label)
- [x] There is **no urgency** with this change and can be released at any time.

## PR Checklist

<!-- Camunda maintains 18 months of versions. Backporting your change to multiple versions is common. -->

- [ ] My changes are for an **already released minor** and are in `/versioned_docs` directory.
- [x] My changes are for the **next minor** and are in `/docs` directory (aka `/next/`).


- [x] I included my new page in the sidebar file(s).
- [ ] I added a redirect for a renamed or deleted page to the .htaccess file.


<!-- All changes require either an Engineering review or technical writer review. **Many require both!** -->

- [x] My changes require an [Engineering review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned the Engineering DRI or delegate.
- [x] My changes require a [technical writer review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned @camunda/tech-writers as a reviewer.

<!-- UNCOMMENT THIS SECTION IF APPLICABLE. Changes to **docs infra**, including updates to workflows and adding new npm packages, must be first discussed via issue or #ask-c8-documentation and linked for context.
- [ ] My changes require a [docs infrastructure review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process). -->

pepopowitz commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,309 @@
+
+The **sort** object specifies which fields of the object should be sorted and whether they are sorted in ascending (ASC) or descending (DESC) order.
+
+The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstItemSortValues_ and _lastItemSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into ...
```
This is really helpful, actually...I was going to ask you a question about these fields today anyway. 👍

pepopowitz replied:
@tmetzke stopping back after attempting to add spec descriptions for these properties, armed with my new understanding -- one other field that isn't mentioned here is `from`, an int32. 

1. Am I correct in thinking this is probably an index to start from? e.g. `from: 11` means start at the 11th item. 
2. If I am correct, is that a zero-based index or 1-based?
3. Do you think we should describe that here, since we covered the other fields?

pepopowitz replied:
Potentially blocking: 

Since these are fields, I think they should be `code` instead of _italic_. @christinaausley can probably confirm, but that's how I interpret the [code blocks section of the style guide](https://confluence.camunda.com/pages/viewpage.action?spaceKey=HAN&title=Camunda+style+guide).

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the `limit`. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned `firstItemSortValues` and `lastItemSortValues` of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes `searchAfter` or `searchBefore`.
```

pepopowitz replied:
Also blocking:

The [properties in the spec](https://github.com/camunda/camunda/blob/main/zeebe/gateway-protocol/src/main/proto/rest-api.yaml#L4345-L4349) are named `firstSortValues` and `lastSortValues`.

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstSortValues_ and _lastSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes _searchAfter_ or _searchBefore_.
```

pepopowitz commented on versioned_docs/version-8.6/apis-tools/camunda-api-rest/specifications/sidebar.js:
```diff
@@ -5,235 +5,205 @@ module.exports = [
-        id: "apis-tools/camunda-api-rest/specifications/get-cluster-topology",
-        label: "Get cluster topology",
-        className: "api-method get",
+        id: "apis-tools/camunda-api-rest/specifications/pin-internal-clock-alpha",
```
non-blocking, no action requested, informative comment: 

Thanks for doing this reordering!

You might be wondering why you had to do it in two places (here, and in the top-level version-8.6-sidbars.json). As tracked in https://github.com/camunda/camunda-docs/issues/3027, this particular sidebar.js file is unused by docusaurus, because numbered versions use a static `.json` file for the top-level sidebars. This file only exists in the 8.6 docs because docusaurus copied it over during versioning, and we never deleted it.

All that to say, any changes in this file will just be ignored. If anything, we should probably delete this file....but that is not your responsibility.

pepopowitz reviewed (approved):
This is great, and I love it. None of my feedback is blocking.

pepopowitz commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,309 @@
+
+| Operator | Syntax                                    | Description                                                                                                                                               |
+| -------- | ----------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| $eq      | field: { "$eq": value }                   | Filter where field is equal to value. Abbreviated form field: value is also allowed.                                                                      |
```
```suggestion
| `$eq`      | `field: { "$eq": value }`                   | Filter where field is equal to value. Abbreviated form field: value is also allowed.                                                                      |
```

Potentially blocking:

Again, this is my interpretation of the style guide, but I think these bits should all be inline code. (Applies also to many lines that follow this one.)

pepopowitz reviewed (commented):
After I completed my review, I noticed a couple other adjustments that I _think_ we should make. I'm not 100% certain though, so I'm marking this as "commented" rather than "request changes," with hopes that a tech writer can straighten me out.

tmetzke commented:
Thanks for the elaborate feedback, @pepopowitz 👍 
I'm Zeebe medic this week so I'll have a look at it next week 🙂

akeller reviewed (changes requested):
I'm a little confused about the audience for this content. It often reads like something that is only relevant for Camundi and should be tucked into the [howtos](https://github.com/camunda/camunda-docs/tree/main/howtos).

Can you help me understand the intention here?

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,309 @@
+
+We use the following error codes and descriptions across our APIs:
+
+- 200 OK
+- 204 No Content
+- 400 Bad Request
+  - Generic error that contains further description in the problem detail.
+- 401 Unauthorized
+  - The client is not authenticated yet.
+  - The client should try again with a modified Authorization header.
+- 403 Forbidden
+  - The client has incorrect or insufficient permissions for the request.
+- 404 Not Found
+- 409 Conflict
+  - The request is trying to modify a resource that is currently not in the right state.
+- 412 Precondition failed
+  - The client should check the cluster status.
+- 429 Rate Limited Exceeded
+  - The client exceeds a defined limit of requests, e.g. Zeebe signaling backpressure due to more requests than the broker can currently process
+- 500 Internal Server Error
+  - Generic error that contains further description in the problem detail.
```
I would recommend turning this into a table.

danielkelemen commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+POST /v2/user-tasks/search
+
+{
+  "filter": {
+    "assignee": {"$eq": "demo"},
+    "candidateGroups": { "$in": ["groupA", "groupB"] },
+    "variables" : {
+      "orderVolume": 10000,
+      "foo": {"$lt": 500},
+      "bar": {"$exists": false}
+    },
+    "$or": [
+      { "priority": {"$eq": "high"}, "dueDate": { "$lt": "<date>" } },
+      { "priority": {"$eq": "low"}, "followUpDate": { "$lt": "<date>" } }
+    ]
```
🔧 No `$or` yet, add multiple conditions.

```suggestion
  "filter": {
    "assignee": {"$eq": "demo"},
    "candidateGroups": { "$in": ["groupA", "groupB"] },
    "variables" : {
      "orderVolume": 10000,
      "foo": {"$gt": 100, "$lt": 500},
      "bar": {"$exists": false}
    }
```

akeller commented:
> @akeller @tmetzke This is a good conversation. I know Tobias wants to version these guidelines, but I'm wondering what alternatives we could consider here.

Who is the audience? Camundi, customers, or both? If we want this to be versioned, but not really available for customers, we can keep it out of the sidebar.

tmetzke commented:
Thanks, everyone, for your input here, it has not gone unnoticed, I'm just too buzzed up with so many other topics that are blocking other people that I couldn't get to this one here 🙈

akeller commented:
@christinaausley let's do a little more lifting for @tmetzke. Based on his response here, I can review this a bit to soften some of the Camundi-only guidance and make this more approachable for Camundi and non-Camundi users. 

Could you also pull in the suggestions? And turn the error codes into a table format?

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+description: "Learn about the basic guidelines, structures, and conventions of the Camunda 8 REST API."
+---
+
+Camunda follows a mix of proposed standards and best practices for RESTful design and consistent implementation across all components. This ensures customers working across all component APIs have a consistent and expected experience without having to study our API reference material or perform “validation testing” to see how our APIs respond.
```
```suggestion
Camunda follows a mix of proposed standards and best practices for RESTful design and consistent implementation across all components. 
```

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+
+## Naming conventions
+
+Naming should be simple, intuitive, and consistent across Camunda 8 APIs. This improves the experience when working across multiple APIs by reducing friction.
```
```suggestion

```

I would remove these sentences or change the tense. "Naming is simple, intuitive, ..."

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+
+Camunda follows a mix of proposed standards and best practices for RESTful design and consistent implementation across all components. This ensures customers working across all component APIs have a consistent and expected experience without having to study our API reference material or perform “validation testing” to see how our APIs respond.
+
+## Naming conventions
```
This section should be rewritten for a non-Camundi audience or consumers of the API, not builders of the API.

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+
+Camunda uses the term “major version number” from [semantic versioning](https://semver.org/), but do not follow semantic versioning for APIs outright. Instead, Camunda provides updates to the API in place and only increment the version number when a major, breaking change happens.
+
+Adding attributes and endpoints are not considered breaking changes. Breaking changes can potentially break an integration. See [GitHub’s REST documentation](https://docs.github.com/en/rest/about-the-rest-api/breaking-changes?apiVersion=2022-11-28#about-breaking-changes-in-the-rest-api) for a comprehensive summary of breaking changes. No migration is required unless there is a breaking change.
```
```suggestion
:::note 
New attributes and endpoints are not considered breaking changes.
:::

Camunda avoids breaking changes that may break integrations or require migrations. Camunda follows [GitHub’s REST documentation](https://docs.github.com/en/rest/about-the-rest-api/breaking-changes?apiVersion=2022-11-28#about-breaking-changes-in-the-rest-api) for a comprehensive summary of breaking changes. 
```

Feel free to further tweak this, but this is the general idea I'd like to see.

akeller commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,307 @@
+
+## HTTP status codes & error handling
+
+Handling errors must be consistent across all endpoints, using well-known HTTP status codes and clear descriptions. This includes the information about errors and the use of a problem details object.
```
"Handling errors is consistent..."

christinaausley commented:
@tmetzke Sorry for my delay here. I will give this a thorough review on Monday.

christinaausley commented:
I have made a handful of formatting and grammatical adjustments -- WDYT?

christinaausley commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md:
```diff
@@ -0,0 +1,338 @@
+The sort object specifies which fields of the object should be sorted and whether they are sorted in ascending (ASC) or descending (DESC) order.
+</TabItem>
+<TabItem value="page">
+The page object details how to slice the result set. An initial search request can omit the page object or define the limit. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned firstItemSortValues and lastItemSortValues of the [search responses](#search-responses) to page through the items by copying that array into one of th ...
```
Did we ever figure out a workaround for linking within tabs?

akeller replied:
@christinaausley do you mean linking from one tab to another? You might be able to use a [query string](https://docusaurus.io/docs/markdown-features/tabs#query-string), assuming the tabs are in the same group (since it's not an anchor, the browser won't scroll).

christinaausley replied:
As in, can we link to another page from within a tab. Nevermind though, as I realized this just looked wonky on my local build and looks fine on deploy 👍

christinaausley commented:
@tmetzke How can I best assist you with this? Are you looking for additional review?

tmetzke commented:
I want to rework this a bit based on previous comments and feedback. I have some ideas but simply get interrupted by too many other topics eventually 😅 
It's getting closer to the top of my list currently, so maybe I can finally get an hour or two in for this to improve it 🙈

tmetzke commented:
Notes:

* Add a hint that data is always "live" and can "shift" if new data comes in
* Add examples for "offset"

tmetzke commented:
Note:
We added more content to naming conventions, namely:

> * key and id fields contain the entity as a prefix, for example, userTaskKey, processDefinitionId. This applies when referencing other resources like formKey in the user task entity, in the respective entities themselves like userTaskKey in the user task entity.
> * The full entity is the prefix to avoid confusion, for example processDefinitionKey instead of processKey (the latter could be interpreted as process instance or process definition).
> * Other attributes of entities themselves have no prefix to avoid clutter, for example version in the process definition entity. In other resources, they have to be referenced with a prefix, like processDefinitionVersion in the process instance entity.

christinaausley commented:
@tmetzke Based on the issue it looks like you recently moved this back into `In progress` -- please let me know if I can assist in any way 😄

tmetzke commented:
Made some progress today, I'll need about another hour to finish my run, will do that tomorrow morning CET.

tmetzke commented:
I'm done here for now. Happy for review feedback 🙂

tmetzke commented:
Some thoughts on the new structure:
* I split it into two parts only, as I find eight (or more) sub-pages harder to consume as a docs user, personally. But maybe I am wrong with this.
* I removed the tabs and tab items. I don't enjoy "jumping" pages when I flip through tabs because of differently-sized tab contents. I like it with different implementation options, meaning the same content but different languages, for example. I don't really enjoy it with completely different content. Again, just my personal preference, maybe I am wrong with this.

akeller commented on docs/apis-tools/camunda-api-rest/guidelines/data-fetching.md:
```diff
@@ -0,0 +1,324 @@
+import Tabs from "@theme/Tabs";
+import TabItem from "@theme/TabItem";
+
+All persistent resources like process definitions, user tasks, users, or tenants provide at least one endpoint to fetch related data. Most of those endpoints provide data with **near-real time consistency** queried from exported records, if records for the respective resource are exported by the engine. If resources are not based on exported records, such as license data or topology information,  ...
```
Is there somewhere else I can learn about persistent resources? Do we cover this anywhere else in the docs?

akeller reviewed (commented):
I'm still struggling with the purpose and audience for this content, but for this review, I find myself asking why we need an "overview" and a conventions page, separately. Could you consider consolidating these? Either in this PR or a future one.

This PR needs updates for the rename to the "Orchestration Cluster API". Right now the content reads like it's for all APIs, which I don't think is correct. 

I also find myself wondering if there are any differences between this API and the other C8 APIs. Do I need to map existing knowledge and understanding, or do I need to learn something entirely different? Or in the migration context, anything new or different than C7 APIs? Can we add more cross links?

tmetzke commented:
I'll look into the first two points within this PR short term to make it more consistent and worthwhile.
For the linking, I might resort to a follow-up issue.

megglos commented:
> @megglos, are there any plans for a migration guide (with or without detailed steps for specific endpoints) from C7 REST API to C8 Orchestration Layer API, similar to (or maybe even becoming a subpage of) the [V1 to V2 API migration guide](https://docs.camunda.io/docs/next/apis-tools/migration-manuals/migrate-to-camunda-api/)?

@berndruecker and team are currently working on this, see https://docs.camunda.io/docs/next/guides/migrating-from-camunda-7/code-conversion/#api-mapping-guide which links to this tool https://camunda-community-hub.github.io/camunda-7-to-8-code-conversion/

tmetzke commented:
> https://docs.camunda.io/docs/next/guides/migrating-from-camunda-7/code-conversion/#api-mapping-guide which links to this tool https://camunda-community-hub.github.io/camunda-7-to-8-code-conversion/

Maybe we should align how the API migration guides look like 😄

megglos commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-data-fetching.md:
```diff
@@ -0,0 +1,324 @@
+
+Many resources offer search endpoints to query by POST method and a given query request object where applicable. The structure of such search requests always follows the same schema and so does the response, always returning a list of items matching the query criteria, accompanied by search result metadata.
+
+[Search requests](#search-requests) are forwarded as queries to the data stores that hold the exported records and the search results are returned in the format described in [search responses](#search-responses).
```
💭 could be shortened or omitted based on same thoughts as previous comment

tmetzke replied:
Yes, I agree 👍

megglos reviewed (approved):
LGTM!

I shared some thoughts on actually not going into much details on matters like exported records and such. For internal guidelines it may fit for further context, for public API user facing docs maybe less. Curious on your and @akeller thoughts!

tmetzke commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-data-fetching.md:
```diff
@@ -0,0 +1,324 @@
+import Tabs from "@theme/Tabs";
+import TabItem from "@theme/TabItem";
+
+All persistent resources like process definitions, user tasks, users, or tenants provide at least one endpoint to fetch related data. Most of those endpoints provide data with **near-real time consistency** queried from exported records, if records for the respective resource are exported by the engine. If resources are not based on exported records, such as license data or topology information,  ...
```
We also have https://github.com/camunda/camunda/issues/26297 on the roadmap that ties in with this. It would highlight those endpoints with near-real-time consistency guarantees.

tmetzke commented:
> @tmetzke could you bring in someone from consulting to review? Maybe someone quite familiar with our APIs and SDKs. I'm curious if they want the level of detail here or reinforce what @megglos and I are honing in on - too much detail for the external/customer audience.
> 
> The other option is for you to make a few changes and simply ship this. We could do UX research on this page and build a backlog of feedback from there.

I am all for keeping this simple for external users and rather linking to pages that provide more detail for users that want to go there. But happy to collect feedback after we ship the simple version.

christinaausley commented:
@tmetzke Does this mean you are ready for a final TW review? 😄

tmetzke commented:
> @tmetzke Does this mean you are ready for a final TW review? 😄

YESSSS! At least from my end 😆

megglos reviewed (dismissed):
👍  One more suggestion on the exporter reference. Alternatively I would also be fine with not mentioning that implementation detail within the API docs.

christinaausley reviewed (dismissed):
Overall, this looks good -- I give it my approval to just ship it as a solid first iteration 👍

tmetzke commented:
@megglos, I need a new approval after the latest commit (fixing the exporter reference) 🤷

megglos reviewed (approved):
🚀

megglos commented on docs/apis-tools/camunda-api-rest/camunda-api-rest-data-fetching.md:
```diff
@@ -0,0 +1,324 @@
+
+Search endpoints can also be used to access entity instances with a unique identifier directly. As an alternative, resources can also provide GET method endpoints for fetching the data of single instances. This is done by a specific key parameter in the URL path, for example `GET /v2/user-tasks/{userTaskKey}`.
+
+All those data retrieval endpoints offer near-real-time data consistency guarantees, providing runtime and historic data that has been processed by the [Camunda Exporter](/self-managed/zeebe-deployment/exporters/camunda-exporter.md`).
```
```suggestion
All those data retrieval endpoints offer near-real-time data consistency guarantees, providing runtime and historic data that has been processed by the [Camunda Exporter](../../self-managed/zeebe-deployment/exporters/camunda-exporter.md`).
```

tmetzke replied:
That's interesting, absolute paths references should work just fine, they work for other links as well 🤔

tmetzke replied:
Yeah, I guess we'd have to define linting in the docs if we wanted to avoid either. Both have their pros and cons, so 🤷
